### PR TITLE
Add `OsStr::display` as a counterpart to `Path::display`

### PR DESCRIPTION
**Pull request by jyn514**
Displaying the string could already be done with `Path::new(os_str).display()`; this just removes the middleman.

I marked this as insta-stable, but I'm happy to change it to unstable if you think it's better to get feedback on nightly.

### CODE REVIEW

File: library/std/src/ffi/os_str.rs [hunk 1]
```diff
@@ -1068,8 +1068,31 @@ impl fmt::Debug for OsStr {
...
+    ///
+    /// let os_str = OsStr::new("foo.txt");
+    /// println!("{}", os_str.display());
+    /// ```
+    ///
+    /// ```
+    /// let path = std::path::Path::new("foo.txt");
+    /// let os_str = path.as_os_str();
+    /// println!("{}", os_str.display());
+    /// ```
```
Comment (m-ou-se) [inline library/std/src/ffi/os_str.rs:1093]: Why this example? For displaying a `Path`, `path.display()` should just be used directly, right?

File: library/std/src/ffi/os_str.rs [hunk 2]
```diff
@@ -1068,8 +1068,31 @@ impl fmt::Debug for OsStr {
 }
 
 impl OsStr {
-    pub(crate) fn display(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
-        fmt::Display::fmt(&self.inner, formatter)
+    #[stable(feature = "os_str_display", since = "1.51.0")]
+    /// Returns an object that implements [`Display`] for safely printing strings that may contain
+    /// non-Unicode data.
+    ///
+    /// See also [`Path::display()`].
```
Comment (m-ou-se) [inline library/std/src/ffi/os_str.rs:1075]: There's not any more information in `Path::display()`'s documentation, right?

File: library/std/src/ffi/os_str.rs (hunk 1 above)
Comment (jyn514) [reply library/std/src/ffi/os_str.rs:1093]: Ok, I changed this to assert that `path.display()` and `path.as_os_str().display()` are the same instead.

File: library/std/src/ffi/os_str.rs (hunk 2 above)
Comment (jyn514) [reply library/std/src/ffi/os_str.rs:1075]: I guess not, no. Should I remove this?

**Comment by m-ou-se** [comment]
It's a bit unfortunate that `OsStr::display()` will return `impl fmt::Display + '_`, while `Path::display()` returns a concrete type (`std::path::Display`). I suppose they could both return the same `Display` struct, but then having it under `std::path` isn't great. We could move it and re-export it from `std::path`. But that might not be worth it.

What do you think?

**Comment by jyn514** [comment]
> It's a bit unfortunate that OsStr::display() will return impl fmt::Display + '_, while Path::display() returns a concrete type

The inconsistency seems unfortunate, but I'm not sure why `Path::display` returns a concrete type in the first place. I don't think you can do anything with `path::Display` other than print it, right?

**Comment by jyn514** [comment]
Also, `std::ffi::Display` seems kind of strange, and making a new module just for this seems like overkill.

**Comment by m-ou-se** [comment]
@rfcbot merge

This adds `OsStr::display()`, the equivalent of `Path::display()` for `OsStr`. It doesn't add a new public type, but it uses `impl Display + '_` as return type.

**Comment by Mark-Simulacrum** [comment]
Is this the first use of a impl trait return type in std?

**Comment by BurntSushi** [comment]
@rfcbot concern impl-trait

I don't think `impl Trait` is suitable here. It is too restrictive. And I don't see any reason why it shouldn't be consistent with Path::display and return its own concrete type. We probably should create a new type, perhaps defined as `pub struct Display<'a>(std::path::Display<'a>)` in the std::ffi module.

In general, I don't think we should use `impl Trait` as the return type in public APIs unless we have to. And I don't think we have to here.

**Comment by jyn514** [comment]
I really wish associated types in impl blocks were implemented. `OsStr::Display` would be a lot more clear than `std::ffi::Display`.

**Comment by BurntSushi** [comment]
Also, does anyone on the library team know if we omitted a Display impl on OsStr on purpose? I can't remember myself. cc @SimonSapin

File: library/std/src/ffi/os_str.rs (hunk 2 above)
Comment (jyn514) [reply library/std/src/ffi/os_str.rs]: I removed the link.

**Comment by jyn514** [comment]
Ok, this now moves `std::path::Display` to `std::ffi` and re-exports it to `std::path` for compatibility.

**Comment by jyn514** [comment]
> Also, does anyone on the library team know if we omitted a Display impl on OsStr on purpose?

I can't imagine why you would - like I said above, you can get the current behavior currently with `Path::new(os_str).display()`.

**Comment by Mark-Simulacrum** [comment]
It seems like we probably want a display on CStr too, then, if we're adding it to OsStr FWIW.

**Comment by BurntSushi** [comment]
> > Also, does anyone on the library team know if we omitted a Display impl on OsStr on purpose?
> 
> I can't imagine why you would - like I said above, you can get the current behavior currently with `Path::new(os_str).display()`.

Yes, I'm aware. I'm doing my due diligence and asking anyway though.

And do we want a re-export here instead of a separate concrete type? I would naturally assume the latter. It is more conservative/flexible, and gives us more freedom later, if necessary.

**Comment by dtolnay** [comment]
@rfcbot concern module
OsStr is not the only current or future thing in std::ffi that one would want a display adapter for, so using std::ffi::Display for this seems inappropriate. Maybe std::ffi::os_str::Display?

**Comment by camelid** [comment]
> > It's a bit unfortunate that OsStr::display() will return impl fmt::Display + '_, while Path::display() returns a concrete type
> 
> The inconsistency seems unfortunate, but I'm not sure why `Path::display` returns a concrete type in the first place. I don't think you can do anything with `path::Display` other than print it, right?

See #80676.

**Comment by jyn514** [comment]
I am probably not going to get around to this. Happy for someone else to pick it up.

**Pull request closed without merging**